src/plans/conjugate_gradient_plan.jl

```diff
@@ -507,3 +508,81 @@ function (u::SteepestDirectionUpdateRule)(
 )
     return 0.0
 end
+
+@doc raw"""
+    ConjugateGradientRestart <: DirectionUpdateRule
+
+An update rule might require a restart, that is using pure gradient as descent direction, if the last two gradients
+are nearly orthogonal, cf. [^HagerZhang2006], page 12 (in the pdf, 46 in Journal page numbers).
+This method acts as a _decorator_ to any existing [`DirectionUpdateRule`](@ref) `direction_update`.
+
+When obtain from the [`ConjugateGradientDescentState`](@ref)` cgs` the last
+``p_k,X_k`` and the current ``p_{k+1},X_{k+1}`` iterate and the gradient, respectively.
+
+Then a restart is performed, i.e. ``β_k = 0`` returned if
+
+```math
+    \frac{ ⟨X_{k+1}, P_{p_{k+1}\gets p_k}X_k⟩}{\lVert X_k \rVert_{p_k}} > ξ,
+```
+where ``P_{a\gets b}(⋅)`` denotes a vector transport from the tangent space at ``a`` to ``b``,
+and ``ξ`` is the `threshold`.
+The default threshold is chosen as `0.2` as recommended in `
+
+# Constructor
+
+    PolakRibiereCoefficient(
+        direction_update::D,
+        threshold=0.2;
+        manifold::AbstractManifold = DefaultManifold(),
+        vector_transport_method::V=default_vector_transport_method(manifold),
+        a::StoreStateAction=StoreStateAction((:Iterate, :gradient, :δ)),
+    )
+
+[^HagerZhang2006]:
+    > W. W. Hager and H. Zhang, A Survey of Nonlinear Conjugate Gradient Methods
+    > Pacific Journal of Optimization 2, 2006, pp. 35-58.
+    > url: [http://www.yokohamapublishers.jp/online2/pjov2-1.html](http://www.yokohamapublishers.jp/online2/pjov2-1.html)
+
+[^Powell1977]:
+    > M.J.D. Powell: Restart Procedures for the Conjugate Gradient Method,
+    > Methematical Programming 12, 1977, pp. 241–254
+    > doi: [10.1007/BF01593790](https://doi.org/10.1007/BF01593790)
+"""
+mutable struct ConjugateGradientRestart{
```

Comment:
Since there are many different restart strategies and this one is very specific, maybe let's call it `ConjugateGradientRestartBeale` (after the name from Powell1977)?

Reply:
why `Beale`?

Reply:
That's what Powell calls this method.

Reply:
Ah I see, but then we should also cite Beale ;)

Reply:
Sure, why not :slightly_smiling_face: .

Reply:
Will try to find the exact reference then later today.

Reply:
Powell refers to this paper:
E.M.L. Beale, "A derivation of conjugate gradients", in: F.A. Lootsma, ed., Numerical methods for nonlinear optimization (Academic Press, London, 1972) pp. 39-43.

Reply:
...yes, but does it have a DOI? Or at least a URL? Will check.

Reply:
Could not find a URL but at least added an ISBN.